### ISSUE TITLE
Named type arguments

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#32540]
When a class has multiple type arguments, it's not always obvious what they each are for. It would be nice if we could name them, the same way we name regular arguments. If we could simultaneously make the named ones optional and give them default values (separate from their lower bound), it would make it significantly easier to then add type arguments without breaking clients of the original API.

**Comment by matanlurey** [comment]
This is interesting, but I wonder if the motivation is more important.

With type inference, you should rarely (?) need to type the argument. For example:

```dart
class Tuple<A, B> {
  final A a;
  final B b;
  Tuple(this.a, this.b);
}

void main() {
  // Automatically infers Tuple<String, int>
  var tuple = new Tuple('Hello', 5);
}
```

What I _would_ like to see, instead, is a default argument. For example:

```dart
class Tuple<A, B, C = Null> { ... }

void main() {
  // Automatically infers Tuple<String, int, Null>.
  var tuple = new Tuple<String, int>();
}
```

Adding a wildcard `?` could also help (similar to `_`):

```dart
class Tuple<A, B = Null, C = Null> { ... }

void main() {
  // Automatically infers Tuple<String, Null, int>.
  var tuple = new Tuple<String, ?, int>();
}
```

... I realize that doesn't solve the problem of if you _want_ to invoke it manually (without inference), or specifying what the arguments are _for_ without looking at the API - but I'd be curious what languages with generics don't use positionals.

**Comment by Hixie** [comment]
I was mostly thinking of generic methods, where the type arguments are often actual arguments (e.g. "search for objects of this type", or "create objects of type Foo with this type as its type argument").

**Comment by matanlurey** [comment]
Ah, that's a good point. To be clear, something like:

```dart
abstract class Context {
  T findWidgetOfType<T>();
}
```

And you'd like to, for example, specify:

```dart
abstract class Context {
  /// NEW: Specify 'P' to filter by widgets that have 'P' as a parent.
  ///   findWidgetOfType<Foo, Bar>();
  ///   ... will find widgets of type 'Foo' where 'Bar' is the direct parent.
  T findWidgetOfType<T, P>();
}
```

... but that would be a breaking change, forcing you to write:

```dart
T findWidgetOfTypeWithParent<T, P>();
```

... correct?

**Comment by Hixie** [comment]
Yeah, that's a great example. Currently, we either have to make a new API or force any users of this API to add a `, dynamic` to the end of any such call. Ideally we'd be able to say `findWidgetOfType<Foo, parent: Bar>` and have `findWidgetOfType<Foo>` imply `parent: dynamic`.

**Comment by Hixie** [comment]
Another example which uses class type arguments would be a class that represents a function call, whether one type is the type of the arguments and one type is the type of the return value. One might imagine wanting to default both to `void` and only requiring that they be specified for instances that actually have an argument or have a return value, respectively. (Specifically, I'm thinking of ModalRoutes in Flutter, where having arguments and having return values are not common, and having both is even less common, but where when you do have them, it's nice to be able to get some extra type checking.)

**Comment by Hixie** [comment]
(Type inference doesn't typically help in those cases, especially for the return value, since the value is provided much later. For the argument it might sometimes help. In any case, right now you always have to specify both if you specify either, which is awkward.)

**Comment by lrhn** [comment]
Sounds like asking for type parameters to have the full complexity of function parameters (optional, named, default values).

While clearly possible, I'd loathe to have `T Function<T, [S extends num]>(T, [S])` be assignable to `T Function<T>(T)`.

Maybe it'd be easier to allow normal parameters to carry types, and not have a separate type parameter syntax (probably not, though, since that would likely mean using `Type` objects as types, unless we define `Type` parameters to be special, like `T foo(T x, {type T}) => x;`. That's probably too complex too.

**Comment by eernstg** [comment]
I guess the point of this issue could be to remember that this particular language design space has a simple end (the current design) and a complex end (@lrhn's full generalization of type parameters); and maybe @Hixie's proposal (allowing some type parameters to have a name and a default value) would be a useful sweet spot between expressive power and counter-productive complexity.

One thing comes to mind, though; with `class A<X, y: Y extends num = int, Z, w: W = void> {...}`, what does `A<double, double, double>` mean, `A<double, int, double, double>` or `A<double, double, double, void>`? ;-)

**Comment by lrhn** [comment]
If we have named type parameters, they should probably *require* a name when applied.

We could omit *optional positional* type parameters, so only named type parameters can be optional.
It seems like a lack of generality, though.

**Comment by Hixie** [comment]
Being able to use Type values as types would be amazing.

I agree that if you name the type parameter, the name should be required on use, just as with regular parameters.
